### ISSUE TITLE
allow string literals as defaults for parameters that extend string

### ISSUE THREAD

**Issue by rob3c** [Microsoft/TypeScript#23718]
**TypeScript Version:**  2.9.0-dev.20180426

**Search Terms:** 
typescript generic type inference default parameter string literal extends string is not assignable to type

**Code**
```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.

// why squigglies with default string literal params?
function foo<S extends string = 'foo1'>( // need 'foo1' here or else string
    s: S = 'foo1' // red squigglies: type '"foo1"' is not assignable to type 'S'?!?
): S {
    return s;
}

const foo1 = foo();       // type "foo1", not string - yay! 
const foo2 = foo('foo2'); // type "foo2", as expected


// The overloads below work without red squigglies, but seem unnecessary
function bar(): 'bar1';
function bar<S extends string>(s: S): S;
function bar(s = 'bar1') {
    return s;
}

const bar1 = bar();       // type "bar1", not string - yay! 
const bar2 = bar('bar2'); // type "bar2", as expected
```

**Expected behavior:**
I expect to be able to assign a string literal as a default value without red squigglies for a function parameter that extends string.

**Actual behavior:**
I get red squigglies complaining at the function declaration (which causes build errors in webpack and such), but it still 'works' by inferring the correct type for the caller.

**Playground Link:**  
https://www.typescriptlang.org/play/#src=%2F%2F%20why%20squigglies%20with%20default%20string%20literal%20params%3F%0D%0Afunction%20foo%3CS%20extends%20string%20%3D%20'foo1'%3E(%20%2F%2F%20need%20'foo1'%20here%20or%20else%20string%0D%0A%20%20%20%20s%3A%20S%20%3D%20'foo1'%20%2F%2F%20red%20squigglies%3A%20type%20'%22foo1%22'%20is%20not%20assignable%20to%20type%20'S'%3F!%3F%0D%0A)%3A%20S%20%7B%0D%0A%20%20%20%20return%20s%3B%0D%0A%7D%0D%0A%0D%0Aconst%20foo1%20%3D%20foo()%3B%20%20%20%20%20%20%20%2F%2F%20type%20%22foo1%22%2C%20not%20string%20-%20yay!%20%0D%0Aconst%20foo2%20%3D%20foo('foo2')%3B%20%2F%2F%20type%20%22foo2%22%2C%20as%20expected%0D%0A%0D%0A%0D%0A%2F%2F%20The%20overloads%20below%20work%20without%20red%20squigglies%2C%20but%20seem%20unnecessary%0D%0Afunction%20bar()%3A%20'bar1'%3B%0D%0Afunction%20bar%3CS%20extends%20string%3E(s%3A%20S)%3A%20S%3B%0D%0Afunction%20bar(s%20%3D%20'bar1')%20%7B%0D%0A%20%20%20%20return%20s%3B%0D%0A%7D%0D%0A%0D%0Aconst%20bar1%20%3D%20bar()%3B%20%20%20%20%20%20%20%2F%2F%20type%20%22bar1%22%2C%20not%20string%20-%20yay!%20%0D%0Aconst%20bar2%20%3D%20bar('bar2')%3B%20%2F%2F%20type%20%22bar2%22%2C%20as%20expected%0D%0A

**Related Issues:**
This seems like the kind of thing that would've been reported before, but I've searched a fair bit and can't seem to find an exact issue about it. It's probably a matter of the keywords being too general, but grouping searches with quotes in google still didn't find anything. Sorry about the likely duplication in advance, but I did give it my best shot searching...

**Comment by rob3c** [comment]
Yes, but `foo<"bar">()` is where the compiler should be showing an error IMO, not at the string literal default assignment, even though that makes you compiler writers unhappy ;-)

I think typescript should help support expressing the intent of the code, and the intent here is obviously to provide a string literal default value while maintaining the string literal type throughout the function and into the return type. The caller is really the one doing the wrong thing intent-wise by contradicting the default, and I'd hope there was a way to help them catch that. The default 'foo1' value could be considered a constraint to prevent a mismatched explicit 'bar' generic parameter without the matching argument. This idiom may not be the best way to do it, but there should be *some* way to do it.

Otherwise, I don't see a straight-forward way to flow multiple string literal types through functions while still allowing strongly-typed string literal default values for some or all of them. And by straight-forward, I mean without a combinatorial explosion of overloads as would happen when extending my second (working) example, assuming you could even write them all to 'behave' correctly just to maintain the right types without downconversion to strings.

This seems like a serious limitation of typescript to me, but I might be missing something obvious. For example, how can I write the equivalent of the following function that maintains the string literal types in the return value while preventing caller generic param errors and also appeasing typescript?

```ts
function flowStringLiteralTypesWithDefaultsBroken<
    S1 extends string,
    S2 extends string,
    S3 extends string
>(
    // these all have red squigglies for the above mentioned reason
    s1: S1 = 'def1',
    s2: S2 = 'def2',
    s3: S3 = 'def3' 
): {
    s1: S1;
    s2: S2;
    s3: S3;
} {
    return { s1, s2, s3 };
}

// how can I force this to be "def1", "def2", "def3" instead of string, string, string?
const try1 = flowStringLiteralTypesWithDefaultsBroken();

// this is "a", string, string instead of the desired "a", "def2", "def3"
const try2 = flowStringLiteralTypesWithDefaultsBroken('a');

// this obviously works, but isn't the interesting case
const try3 = flowStringLiteralTypesWithDefaultsBroken('a', 'b', 'c');
```

**Comment by RyanCavanaugh** [comment]
> Yes, but `foo<"bar">()` is where the compiler should be showing an error IMO

Let's stop right here - *why and how*? The default value of the parameter is not visible to external callers; it is not part of the type and is not present in `.d.ts` files.

> I think typescript should help support expressing the intent of the code

Overloads?
```ts
function f(): "foo1";
function f<T extends string>(s: T): T;
function f(.. // implementation signature
````

**Comment by rob3c** [comment]
slightly larger example where overloads start getting fun
```ts
function keepTheStringLiteralsWithDefaults<
    S1 extends string = 'def1',
    S2 extends string = 'def2',
    S3 extends string = 'def3',
    S4 extends string = 'def4',
    S5 extends string = 'def5',
    S6 extends string = 'def6',
    S7 extends string = 'def7'
>({
    s1 = 'def1',
    s2 = 'def2',
    s3 = 'def3',
    s4 = 'def4',
    s5 = 'def5',
    s6 = 'def6',
    s7 = 'def7'
}: {
    s1?: S1;
    s2?: S2;
    s3?: S3;
    s4?: S4;
    s5?: S5;
    s6?: S6;
    s7?: S7;
}): {
    s1: S1;
    s2: S2;
    s3: S3;
    s4: S4;
    s5: S5;
    s6: S6;
    s7: S7;
} {
    return {
        s1: s1 as S1,
        s2: s2 as S2,
        s3: s3 as S3,
        s4: s4 as S4,
        s5: s5 as S5,
        s6: s6 as S6,
        s7: s7 as S7,
    };
}

const aok = keepTheStringLiteralsWithDefaults<'a'>({ s1: 'a' });
const aoops = keepTheStringLiteralsWithDefaults<'a', 'y', 'z'>({ s1: 'a' });
```

**Comment by jack-williams** [comment]
Is this more tractable? I think that the lack of object literal inference can make it slightly ugly
```ts
type Keys = keyof Defaults;

interface Defaults {
  s1: "def1";
  s2: "def2";
  s3: "def3";
  s4: "def4";
  s5: "def5";
  s6: "def6";
  s7: "def7";
}

const defaults: Defaults; // ....

type Populate<T extends Partial<Record<Keys, string>>> =
  { [K in keyof Defaults]: undefined extends T[K]  ? Defaults[K] : T[K]};

function populate<R extends Partial<Record<Keys, string>>>(m: R): Populate<R> {
  return Object.assign(defaults, m) as any as Populate<R>;
}

const a: { s1: 'hello' } = { s1: "hello" };
const example = populate(a);
```

**Comment by rob3c** [comment]
thanks @jack-williams, but I'm already familiar with that 'static' approach to the problem. It's ok for the toy example I posted, but that's an artifact of simplification. I tried to simplify the code here to highlight the issue, but unfortunately that also tempts people to offer helpful alternatives that address the toy problem without addressing the complexity of the real problem in actual code.

In any case, the bot closed it, which is fine, and I've chosen to attack my particular use case in a different way until future improvements to typescript fix this. Thanks, everyone!
